delivery_carrier_state/models/stock_picking.py

```diff
@@ -0,0 +1,51 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
+from odoo import fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    date_shipped = fields.Date(
+        string='Shipment Date',
+        readonly=True,
+    )
+    date_delivered = fields.Datetime(
+        string='Delivery Date',
+        readonly=True,
+    )
+    tracking_state = fields.Char(
+        string='Tracking state',
+        readonly=True,
+        index=True,
+    )
+    tracking_state_history = fields.Text(
+        string='Tracking state history',
+        readonly=True,
+    )
+    delivery_state = fields.Selection(
+        selection=[
+            ('shipping_recorded_in_carrier', 'shipping recorded in carrier'),
+            ('in_transit', 'In transit'),
+            ('canceled_shipment', 'Canceled shipment'),
+            ('incidence', 'incidence'),
+            ('customer_delivered', 'Customer delivered'),
+            ('warehouse_delivered', 'Warehouse delivered'),
+        ],
+        string='Carrier State',
+        track_visibility='onchange',
+        readonly=True,
+    )
+
+    def tracking_state_update(self):
+        for picking in self.filtered(lambda p: p.delivery_type):
+            method = '%s_tracking_state_update' % picking.delivery_type
+            if hasattr(picking.carrier_id, method):
+                getattr(picking.carrier_id, method)(picking)
+
+    def cancel_shipment(self, picking):
```

Comment:
Why not doing it in the `cancel_shipment` method of  `delivery.carrier`? It should return the recordset of cancelled pickings: https://github.com/odoo/odoo/blob/12.0/addons/delivery/models/delivery_carrier.py#L158.

Indeed, you could have a delivery that can't be cancelled (i.e.: it's already delivered) wich state shouldn't be changed.

Reply:
You're right, I change it